File: Data Structures & Algorithms/two-integer-sum/submission-2.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        """
        sort + binary search
        """
        indexed_nums = sorted([(n, i) for i, n in enumerate(nums)])
        for i in range(len(indexed_nums)):
            idx = self.bs(indexed_nums, target - indexed_nums[i][0], i + 1)
            if idx != -1:
                return sorted([indexed_nums[i][1], idx])
        return []
    
    def bs(self, nums: List[tuple], target: int, start: int) -> int:
        i = start
        j = len(nums) - 1
        while i <= j:
            m = i + (j - i) // 2
            if nums[m][0] == target:
                return nums[m][1]
            elif nums[m][0] < target:
                i = m + 1
            else:
                j = m - 1
        return -1
        

File: Data Structures & Algorithms/three-integer-sum/submission-2.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ans;
        ans.reserve(nums.size());
        std::sort(nums.begin(), nums.end());
        for (size_t i = 0; i < nums.size() - 2; ++i) {
            if (i > 0 && nums[i] == nums[i-1])
                continue;
            twoSum(nums, i, -nums[i], ans);
        }
        return ans;
    }

    void twoSum(const vector<int>& nums, const size_t pivot, const int target, vector<vector<int>>& ans) {
        size_t start = pivot + 1;
        size_t end = nums.size() - 1;
        while (start < end) {
            if (nums[start] + nums[end] == target) {
                ans.push_back({nums[pivot], nums[start], nums[end]});
                while (start < end && nums[start] == nums[start + 1]) {
                    ++start;
                }
                while (start < end && nums[end] == nums[end - 1]) {
                    --end;
                }
                ++start;
                --end;
            } else if (nums[start] + nums[end] > target) {
                --end;
            } else {
                ++start;
            }
        }
    }
};
